lib/datasource.js

```diff
@@ -538,24 +525,112 @@ DataSource.prototype.defineRelations = function(modelClass, relations) {
       if ((targetModel && !isModelDataSourceAttached(targetModel)) ||
           (throughModel && !isModelDataSourceAttached(throughModel))) {
         // Create a listener to defer the relation set up
-        createListener(rn, r, targetModel, throughModel);
+        deferRelationSetup(relationName, r, targetModel, throughModel);
       } else {
         // The target model is resolved
         var params = traverse(r).clone();
-        params.as = rn;
-        params.model = polymorphicName || targetModel;
+        params.as = relationName;
+        params.model = targetModel;
         if (throughModel) {
           params.through = throughModel;
         }
-        modelClass[r.type].call(modelClass, rn, params);
+        modelClass[r.type].call(modelClass, relationName, params);
       }
     });
   }
 };
 
+function validateRelation(relationName, relation) {
+  var rn = relationName;
+  var r = relation;
+  var code;
+
+  // allows for method injection via relation definition in unit tests
+  getRelationValidationMsg = r.getRelationValidationMsg || getRelationValidationMsg;
+
```

Comment:
@ebarault I see and I think I got your point, my concern is, the real [isValidRelationName](https://github.com/strongloop/loopback-datasource-juggler/pull/1301/files#diff-9e9565730a5d5c0bdf028ccde9dd3a4eR599) function is not tested.

For example, if some user changes the switch case code(case 'A_NEW_NAME'), but forget to change the `code` defined in function relationValidate, our test will still pass.